__tests__/integration/query-typings.test.ts

```diff
@@ -12,19 +15,67 @@ describe.each`
   ${"query"}
   ${"paginate"}
 `("$method typings", ({ method }: { method: string }) => {
-  it("allows customers to use their own types", async () => {
-    expect.assertions(1);
-    // added in a junk property that is not part of QueryValue
-    type MyType = { x: number; t: QueryCheckError };
+  it("allows customers to use their own types in queries", async () => {
+    const query = fql`{ "x": 123 }`;
+    const paginatedQuery = fql`[{ "x": 123}].toSet()`;
+
     if ("query" === method) {
-      const result = (await client.query<MyType>(fql`{ "x": 123}`)).data;
+      const result = (await client.query<MyType>(query)).data;
       expect(result).toEqual({ x: 123 });
     } else {
-      for await (const page of client.paginate<MyType>(fql`{ "x": 123}`)) {
+      expect.assertions(2);
+      for await (const page of client.paginate<MyType>(paginatedQuery)) {
+        for (const result of page) {
+          expect(result).toEqual({ x: 123 });
+        }
+      }
+
+      // It is also allowed to provide a query that does not return a page.
+      // When this happenes, the driver treats the result as if a page with
+      // exactly one item is returned.
+      for await (const page of client.paginate<MyType>(query)) {
         for (const result of page) {
           expect(result).toEqual({ x: 123 });
         }
       }
     }
   });
+
+  it("allows customers to infer their own types in queries from fql statements", async () => {
+    const query = fql<MyType>`{ "x": 123 }`;
+    const paginatedQuery = fql<Page<MyType>>`[{ "x": 123}].toSet()`;
+
+    if ("query" === method) {
+      const result = (await client.query(query)).data;
+      expect(result).toEqual({ x: 123 });
+    } else {
+      expect.assertions(2);
+      for await (const page of client.paginate(paginatedQuery)) {
+        for (const result of page) {
+          expect(result).toEqual({ x: 123 });
+        }
+      }
+
+      // It is also allowed to provide a query that does not return a page.
+      // When this happenes, the driver treats the result as if a page with
+      // exactly one item is returned.
+      for await (const page of client.paginate(query)) {
+        for (const result of page) {
+          expect(result).toEqual({ x: 123 });
+        }
+      }
+    }
+  });
+
+  it("allows customers to use subtyped queries", async () => {
+    const query = fql<string>`"hello"`;
+
+    const result = (await client.query<string | number>(query)).data;
+    expect(result).toEqual("hello");
+
+    // And make sure that the opposite is not possible
+    const query2 = fql<string | number>`"hello"`;
+    // @ts-expect-error Argument of type 'Query<string | number>' is not assignable to parameter of type 'Query<string>'.
+    await client.query<string>(query2);
```

Comment:
wow cool assertion here.

Reply:
Would you mind adding how to use this to the development section of the README?

As we tweak type arguments in future work capturing that we can sanity check our types using this and proper linter settings will help out future devs.

For example, we got some work we'd like to do to make less boilerplate for users when definining their own types and when using `.query` calls that return pages (as opposed to using the clients pagination utils - there's cases you might do that in an API for example). We have some of these type improvements documented internally here: https://faunadb.atlassian.net/wiki/spaces/DX/pages/3852697604/Developer+Tooling+Pain#API-Pagination

So this test pattern is something we will likely need aagain.